### DISCUSSION TITLE
【提案】安全解构运算符（...?）

### DISCUSSION THREAD

**Discussion by LeiJinpeng** [JSCIG/es-discuss#68] [Ideas]
使用扩展运算符 `...` 解构非可迭代对象会导致 TypeError。

引入新的运算符 `...?` 当应用于非可迭代对象时，将忽略该对象而不是抛出异常。

```javascript
[0, ...null, 3] // TypeError: null is not iterable
[0, ...?null, 3] // [0, 3]
```

如果 `expr` 是可迭代的，`...?expr` 行为与 `...expr` 相同。

如果 `expr` 不可迭代，`...?expr` 将忽略该对象。

这样无需额外检查 `expr` 是否可迭代，使代码更加简洁清晰。

**Comment by ljharb** [comment]
in this case, `[].concat(0, null || [], 3)` seems like the more idiomatic approach anyways. Is there a use case beyond arrays?
